### ISSUE TITLE
No namespace set on NFD subchart with helm

### ISSUE THREAD

**Issue by linkvt** [NVIDIA/gpu-operator#56]
There is no namespace set on the node-feature-discovery resources when using helm which is a breaking change for me since v1.0.0, the namespace is set correctly for all gpu-operator resources.

IMO the namespace should be at least reused from the main chart but setting (as before) the namespace to `node-feature-discovery` would be the best I think.

In the following output you can see that the namespace is only set for gpu operator namespaces, there is no namespace set for NFD resources:

<details>

```
$ helm template nvidia/gpu-operator --version=1.1.0 --namespace=gpu-operator --include-crds | grep -B4 -A8 "chart: node-feature"
kind: ServiceAccount
metadata:
  name: RELEASE-NAME-node-feature-discovery
  labels:
    helm.sh/chart: node-feature-discovery-0.1.0
    app.kubernetes.io/name: node-feature-discovery
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/version: "v0.5.0"
    app.kubernetes.io/managed-by: Helm
---
# Source: gpu-operator/templates/serviceaccount.yaml
apiVersion: v1
kind: ServiceAccount
--
kind: ClusterRole
metadata:
  name: RELEASE-NAME-node-feature-discovery
  labels:
    helm.sh/chart: node-feature-discovery-0.1.0
    app.kubernetes.io/name: node-feature-discovery
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/version: "v0.5.0"
    app.kubernetes.io/managed-by: Helm
rules:
- apiGroups:
  - ""
  resources:
--
kind: ClusterRoleBinding
metadata:
  name: RELEASE-NAME-node-feature-discovery
  labels:
    helm.sh/chart: node-feature-discovery-0.1.0
    app.kubernetes.io/name: node-feature-discovery
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/version: "v0.5.0"
    app.kubernetes.io/managed-by: Helm
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: RELEASE-NAME-node-feature-discovery
--
kind: Service
metadata:
  name: RELEASE-NAME-node-feature-discovery-master
  labels:
    helm.sh/chart: node-feature-discovery-0.1.0
    app.kubernetes.io/name: node-feature-discovery
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/version: "v0.5.0"
    app.kubernetes.io/managed-by: Helm
    role: master
spec:
  type: ClusterIP
  ports:
--
kind: DaemonSet
metadata:
  name:  RELEASE-NAME-node-feature-discovery-worker
  labels:
    helm.sh/chart: node-feature-discovery-0.1.0
    app.kubernetes.io/name: node-feature-discovery
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/version: "v0.5.0"
    app.kubernetes.io/managed-by: Helm
    role: worker
spec:
  selector:
    matchLabels:
--
kind: Deployment
metadata:
  name:  RELEASE-NAME-node-feature-discovery-master
  labels:
    helm.sh/chart: node-feature-discovery-0.1.0
    app.kubernetes.io/name: node-feature-discovery
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/version: "v0.5.0"
    app.kubernetes.io/managed-by: Helm
    role: master
spec:
  replicas: 1
  selector:
--
kind: Pod
metadata:
  name: "RELEASE-NAME-node-feature-discovery-test-master-connection"
  labels:
    helm.sh/chart: node-feature-discovery-0.1.0
    app.kubernetes.io/name: node-feature-discovery
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/version: "v0.5.0"
    app.kubernetes.io/managed-by: Helm
  annotations:
    "helm.sh/hook": test-success
spec:
  containers:
```

</details>

**Comment by RenaudWasTaken** [comment]
Actually with helm v3 the gpu-operator chart is wrong and the subchart is right !
Unfortunately in helm v3 you are expected to create the namespace and specify the namespace in the helm command...

I'm not a big fan of this breaking change but IIRC there were good reasons related to permissions and storing data in the namespace.

I'll try to get the chart up to date, thanks for reporting bugs!

**Comment by linkvt** [comment]
I think we missunderstood, what you mean is the namespace creation, what I mean is the namespace of the resources like the node-feature-discovery master Deployment.

If you use my command from above without the `grep` part you see that all gpu-operator resources have `metadata.namespace` correctly set to the parameter I pass with `--namespace` to helm - except the node-feature-discovery stuff (which is wrong IMO).
I guess the namespace is not passed correctly to the subchart.

**Comment by shivamerla** [comment]
@linkvt When we deploy using helm3, all gpu+nfd resources are created in the namespace provided during install and gpu-operator managed resources itself are created in **gpu-operator-resources** namespace. Please try the latest chart and let us know if we can close this issue.

**helm install gpu-operator nvidia/gpu-operator -n operator --wait**
```
ubuntu@ip-172-31-34-83:~$ kubectl get all -n operator
NAME                                                              READY   STATUS    RESTARTS   AGE
pod/gpu-operator-6b4cc48f9f-p7cbn                                 1/1     Running   2          19h
pod/gpu-operator-node-feature-discovery-master-5b84b597f4-gfdfx   1/1     Running   1          19h
pod/gpu-operator-node-feature-discovery-worker-8lfl5              1/1     Running   2          19h

NAME                                                 TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)    AGE
service/gpu-operator-node-feature-discovery-master   ClusterIP   10.102.145.156   <none>        8080/TCP   19h

NAME                                                        DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
daemonset.apps/gpu-operator-node-feature-discovery-worker   1         1         1       1            1           <none>          19h

NAME                                                         READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/gpu-operator                                 1/1     1            1           19h
deployment.apps/gpu-operator-node-feature-discovery-master   1/1     1            1           19h

NAME                                                                    DESIRED   CURRENT   READY   AGE
replicaset.apps/gpu-operator-6b4cc48f9f                                 1         1         1       19h
replicaset.apps/gpu-operator-node-feature-discovery-master-5b84b597f4   1         1         1       19h
```

**gpu-operator managed resources in gpu-operator-resources namespace**

```
ubuntu@ip-172-31-34-83:~$ kubectl get all -n gpu-operator-resources
NAME                                           READY   STATUS             RESTARTS   AGE
pod/nvidia-container-toolkit-daemonset-dpl7h   1/1     Running            1          19h
pod/nvidia-dcgm-exporter-mn64s                 0/1     Running   1         19h
pod/nvidia-device-plugin-daemonset-f8hkc       0/1     Running   1         19h
pod/nvidia-device-plugin-validation            0/1     Completed          0          19h
pod/nvidia-driver-daemonset-lc5pb              0/1     Running   1        19h
pod/nvidia-driver-validation                   0/1     Completed          0          19h

NAME                           TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
service/nvidia-dcgm-exporter   ClusterIP   10.106.139.18   <none>        9400/TCP   19h

NAME                                                DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR                                      AGE
daemonset.apps/nvidia-container-toolkit-daemonset   1         1         1       1            1           feature.node.kubernetes.io/pci-10de.present=true   19h
daemonset.apps/nvidia-dcgm-exporter                 1         1         0       1            0           feature.node.kubernetes.io/pci-10de.present=true   19h
daemonset.apps/nvidia-device-plugin-daemonset       1         1         0       1            0           feature.node.kubernetes.io/pci-10de.present=true   19h
daemonset.apps/nvidia-driver-daemonset              1         1         0       1            0           feature.node.kubernetes.io/pci-10de.present=true   19h
```

**Comment by linkvt** [comment]
I digged into this a bit again as I saw the same issue in 1.1.7 (just try the command from above) and it seems that it appears due to the confusing helm behaviour when specifying the `--namespace` parameter, especially to `helm template`, see e.g. https://github.com/helm/helm/issues/3553#issuecomment-521424132 .
`--namespace` is only effective when `{{ .Release.Namespace }}` is used AND when applying to the k8s cluster - it doesn't set `namespace: ...` on every object as you might expect for `helm template`.

Also before https://github.com/NVIDIA/gpu-operator/commit/8501fd7929234855b60975e7c9902f536c82d920 the namespace parameter was specified in some locations making this a bit more confusing. It seems now that the `helm` way is to have an extra tool patching the namespace on every source as it is an `computationally expensive ordeal` (quote) for helm.

@shivamerla thanks for taking another look at this, closing as this is no gpu-operator issue!
